servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/JavaNetSoTimeoutHttpConnectionFilter.java

```diff
@@ -167,24 +174,118 @@ public Single<StreamingHttpResponse> request(final StreamingHttpRequest request)
+            this.timeoutExecutor = timeoutExecutor;
+        }
+
+        @Override
+        public void onSubscribe(Cancellable cancellable) {
+            delegate.onSubscribe(() -> {
+                once();
+                try {
+                    timeoutCancellable.cancel();
+                } finally {
+                    requestCancellable.cancel();
+                }
+            });
+            requestCancellable.delayedCancellable(cancellable);
+            // We only initialize the timer after the successful call to `delegate.onSubscribe(..)` to avoid
+            // RS spec violations from the timer firing before an `onSubscribe(..)` call. This also means we
+            // don't need to worry about `delegate.onSubscribe(..)` throwing since we don't initialize resources
+            // until after the call is successful.
+            timeoutCancellable.delayedCancellable(requestComplete.concat(Completable.never()
+                    .timeout(timeout, timeoutExecutor)).beforeOnError(this::handleInterruptions).subscribe());
+        }
+
+        private void handleInterruptions(Throwable t) {
+            if (once()) {
+                try {
+                    requestCancellable.cancel();
+                } catch (Throwable tt) {
+                    addSuppressed(t, tt);
+                } finally {
+                    Throwable result = t;
+                    // We can get a SocketTimeoutException waiting for a 100 Continue response.
+                    if (t instanceof TimeoutException) {
+                        result = newStacklessSocketTimeoutException("Read timed out after " + timeout.toMillis() +
+                                "ms waiting for response meta-data").initCause(t);
+                    }
+                    delegate.onError(result);
+                }
+            }
+        }
+
+        @Override
+        public void onSuccess(@Nullable StreamingHttpResponse result) {
+            if (once()) {
+                try {
+                    timeoutCancellable.cancel();
+                    if (result != null) {
+                        result = result.transformMessageBody(p -> p.timeout(timeout, timeoutExecutor)
+                                .onErrorMap(TimeoutException.class, t -> newStacklessSocketTimeoutException(
+                                        "Read timed out after " + timeout.toMillis() +
+                                                "ms waiting for the next response payload body chunk")
+                                        .initCause(t)));
+                    }
+                    delegate.onSuccess(result);
+                    return;
+                } catch (Throwable t) {
+                    delegate.onError(t);
+                }
+            }
+            // If we get here the message was not consumed, so we need to clean it up.
+            if (result != null) {
+                toSource(result.messageBody()).subscribe(CancelImmediatelySubscriber.INSTANCE);
```

Comment:
Taking into account that `onSuccess`/`onError` can throw, we should guarantee that cleanup is invoked.
We should probably do the same for `AbstractTimeoutHttpFilter.CleanupSubscriber` in case:
```java
                // We win.
                delegate.onSuccess(result);
```